### PR TITLE
UX Optional Password & Captcha for External Registration

### PR DESCRIPTION
**Pull request by coolaj86**
Re: https://github.com/go-gitea/gitea/issues/4226 and https://github.com/go-gitea/gitea/issues/3837

I see the primary purposes of external accounts as to increase security and convenience (which always go hand-in-hand).

If I have enabled external accounts, I should not require the user to weaken security by forcing them to create a (likely re-used) password.

Likewise, since I've already hand-picked the external account providers that I trust and want to allow, I should not reduce convenience by requiring a secondary captcha.

* [x] Do not require password for external accounts via `REQUIRE_EXTERNAL_REGISTRATION_PASSWORD`
* [x] Do not require password for external accounts via `REQUIRE_EXTERNAL_REGISTRATION_CAPTCHA`
* [x] Only reject omitted passwords on forms that use a password
* [x] Do not provide the option for a password when `ALLOW_ONLY_EXTERNAL_REGISTRATION`

Some other fixes that I want to get in, but may deserve separate PRs:

* [x] Do not require logging in twice when creating account (PR #5033)
* [x] Allow resetting (creating) password via email, even when logged in (PR https://github.com/go-gitea/gitea/pull/5034, Issue https://github.com/go-gitea/gitea/issues/5008)

If you'd like to test it out
====

My pull request is against `master`, but [I run it](https://git.coolaj86.com) as backport to v1.5.1 (includes https://github.com/go-gitea/gitea/pull/5006, https://github.com/go-gitea/gitea/pull/5029, https://github.com/go-gitea/gitea/pull/5033, https://github.com/go-gitea/gitea/pull/5034):

```bash
git clone https://github.com/coolaj86/gitea.git gitea.coolaj86 -b v1.5.1-coolaj86
pushd gitea.coolaj86
TAGS="bindata sqlite" make generate all
```

I would not recommend replacing your existing gitea, but rather creating a symlink so that you can easily switch back if you don't like it. For example, if you keep `gitea` in `/opt/gitea/bin`:

```bash
rsync -av ./gitea /opt/gitea/bin/gitea-v1.5.1-coolaj86
pushd /opt/gitea/bin
mv gitea gitea-v1.5.1
ln -s gitea-v1.5.1-coolaj86 gitea
```

I've run a couple of manual tests so far, so I feel comfortable with someone else trying it out. I won't be pushing any additional changes to that branch (such as the upcoming changes to address the empty checkboxes in the issue) until I've tested them in production for myself.

### CODE REVIEW

**Comment by lafriks** [comment]
What is reason `REQUIRE_EXTERNAL_REGISTRATION_CAPTCHA` why no just reuse captcha setting that is already present?

**Comment by coolaj86** [comment]
My view is that Captcha is the kind of thing you implement for local logins, not the kind of thing you enforce on delegated logins that you already trust, which already do that type of validation.

For my instance (and I think most people would feel this way), I want the greatest protection from bots with the least friction for new users and I think local-only captcha strikes the right balance.

**Comment by strk** [comment]
Is "OpenID" considered "External Registration" ?
Sorry but I'm a bit confused about the state of oAuth2 vs. OpenID(2)
implementations

**Comment by coolaj86** [comment]
TL;DR
----

* OpenID(1): No change introduced by this PR
* OpenID(2)/OIDC: Same changes as OAuth2

Explanation
-----

I haven't changed the behavior of OpenID(1) and it follows a different flow from OAuth2 in the code.

As such I don't think that these changes affect the OpenID(1) flow. I don't want to pursue changing the OpenID(1) flow as part of this PR (it was never very well adopted and doesn't provide a lot of value in my opinion), but I'd be open to updating it as well later on.

OpenID(2) aka OIDC aka OpenID Connect is actually not related to OpenID in any way. It's a strict subset of OAuth2 that _actually_ defines important things like data format (JSON vs urlencoded), endpoint URLs, etc. It's not very well designed for personal use, in my opinion, so it doesn't really meet the original goal of OpenID (IndieAuth is probably better for that), but it is far better than OAuth2 for enterprise-style use and maintains perfect OAuth2 compatibility (which isn't hard seeing how OAuth2 doesn't define any of those things so there's nothing much to be "compatible" with other than the generic concept of "there's some handshakes").

**Comment by nickho** [comment]
Thanks for working on this @solderjs !
I tested your PR on a setup with :
- local VM on Ubuntu and gitea on https
- our corporate AzureAD as an OIDC provider

The workflow is working fine for me. Clicking the link "Connect With OpenId Connect" on the login page redirect to our AzureAD that redirect to the "Add Account Recovery Info". If i enter a username, then everything is working fine inside Gitea.

I do confirm ALLOW_ONLY_EXTERNAL_REGISTRATION is working, but you need to also to put DISABLE_REGISTRATION to false.

gitea-app-ini

I think something is missaligned on the UI but is it an issue with #5006 ?

add-account-recovery

Keep up the good work.

**Comment by strk** [comment]
Is this work still ongoing ? I'd love to see OpenID-Connect with discovery enabled on try.gitea.io ...

**Comment by lunny** [comment]
Please resolve the conflict

**Comment by zeripath** [comment]
@lunny @solderjs conflict resolved.

File: modules/setting/service.go [hunk 1]
```diff
@@ -61,6 +63,8 @@ func newService() {
 	Service.EnableReverseProxyAutoRegister = sec.Key("ENABLE_REVERSE_PROXY_AUTO_REGISTRATION").MustBool()
 	Service.EnableReverseProxyEmail = sec.Key("ENABLE_REVERSE_PROXY_EMAIL").MustBool()
 	Service.EnableCaptcha = sec.Key("ENABLE_CAPTCHA").MustBool(false)
+	Service.RequireExternalRegistrationCaptcha = sec.Key("REQUIRE_EXTERNAL_REGISTRATION_CAPTCHA").MustBool()
```
Comment (techknowlogick) [inline modules/setting/service.go:66]:
```suggestion
	Service.RequireExternalRegistrationCaptcha = sec.Key("REQUIRE_EXTERNAL_REGISTRATION_CAPTCHA").MustBool(Service.EnableCaptcha)
```

**Comment by lunny** [comment]
replaced by #6606
